### PR TITLE
Fixes to launch Magento with PHP 8.1 - part 1

### PR DESCRIPTION
**Pull request by karyna-t**
### Description (*)
Fixes for compatibility with PHP 8.1 (to be able to launch Magento)
Part 1

### Related Issues
1. magento/magento2#34505

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by xmav** [comment]
@mageto run Static Tests, Semantic Version Checker

**Comment by karyna-tsymbal-atwix** [comment]
@mageto run Static Tests, Semantic Version Checker

**Comment by andrewbess** [comment]
@mageto run Static Tests, Semantic Version Checker

File: app/code/Magento/Analytics/Model/ReportXml/ModuleIterator.php [hunk 1]
```diff
@@ -34,7 +34,7 @@ public function __construct(
      *
      * @return array
      */
-    public function current()
+    public function current(): array
```
Comment (ihor-sviziev) [inline app/code/Magento/Analytics/Model/ReportXml/ModuleIterator.php]: this is a backward-incompatible change.
I think we need to add `#[ReturnTypeWillChange]` instead, like that:
```suggestion
    #[\ReturnTypeWillChange]
    public function current()
```

File: app/code/Magento/Analytics/ReportXml/Query.php [hunk 2]
```diff
@@ -77,12 +77,9 @@ public function getConfig()
...
 
     /**
-     * Specify data which should be serialized to JSON
-     * @link http://php.net/manual/en/jsonserializable.jsonserialize.php
-     * @return mixed data which can be serialized by <b>json_encode</b>,
-     * which is a value of any type other than a resource.
+     * @inheritDoc
      */
-    public function jsonSerialize()
+    public function jsonSerialize(): array
```
Comment (ihor-sviziev) [inline app/code/Magento/Analytics/ReportXml/Query.php]:
```suggestion
    #[\ReturnTypeWillChange]
    public function jsonSerialize()
```

File: app/code/Magento/Customer/Model/Indexer/Source.php [hunk 3]
```diff
@@ -82,7 +82,7 @@ public function addFieldToFilter($attribute, $condition = null)
     /**
      * @inheritdoc
      */
-    public function count()
+    public function count(): int
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Indexer/Source.php]:
```suggestion
    #[\ReturnTypeWillChange]
    public function count()
```

File: app/code/Magento/Customer/Model/Indexer/Source.php [hunk 4]
```diff
@@ -92,7 +92,7 @@ public function count()
      *
      * @return Traversable
      */
-    public function getIterator()
+    public function getIterator(): Traversable
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Model/Indexer/Source.php]:
```suggestion
    #[\ReturnTypeWillChange]
    public function getIterator()
```

File: app/code/Magento/Customer/Ui/Component/MassAction/Group/Options.php [hunk 5]
```diff
@@ -81,7 +81,7 @@ public function __construct(
      *
      * @return array
      */
-    public function jsonSerialize()
+    public function jsonSerialize(): array
```
Comment (ihor-sviziev) [inline app/code/Magento/Customer/Ui/Component/MassAction/Group/Options.php]:
```suggestion
    #[\ReturnTypeWillChange]
    public function jsonSerialize()
```

File: app/code/Magento/Ui/Model/Export/SearchResultIterator.php [hunk 6]
```diff
@@ -40,23 +42,23 @@ public function key()
     /**
      * @return void
      */
-    public function next()
+    public function next(): void
```
Comment (ihor-sviziev) [inline app/code/Magento/Ui/Model/Export/SearchResultIterator.php]:
```suggestion
    #[\ReturnTypeWillChange]
    public function next()
```

File: app/code/Magento/Ui/Model/Export/SearchResultIterator.php [hunk 7]
```diff
@@ -40,23 +42,23 @@ public function key()
...
     {
         next($this->items);
     }
 
     /**
      * @return void
      */
-    public function rewind()
+    public function rewind(): void
```
Comment (ihor-sviziev) [inline app/code/Magento/Ui/Model/Export/SearchResultIterator.php]:
```suggestion
    #[\ReturnTypeWillChange]
    public function rewind()
```

File: app/code/Magento/Ui/Model/Export/SearchResultIterator.php [hunk 8]
```diff
@@ -40,23 +42,23 @@ public function key()
...
     {
         reset($this->items);
     }
 
     /**
      * @return bool
      */
-    public function valid()
+    public function valid(): bool
```
Comment (ihor-sviziev) [inline app/code/Magento/Ui/Model/Export/SearchResultIterator.php]:
```suggestion
    #[\ReturnTypeWillChange]
    public function valid()
```

File: app/code/Magento/Shipping/Model/Simplexml/Element.php [hunk 9]
```diff
@@ -32,9 +32,9 @@ public function addAttribute($name, $value = null, $namespace = null)
      * @param string $name The name of the child element to add.
      * @param string $value If specified, the value of the child element.
      * @param string $namespace If specified, the namespace to which the child element belongs.
-     * @return \Magento\Shipping\Model\Simplexml\Element
+     * @return \Magento\Shipping\Model\Simplexml\Element|null
      */
-    public function addChild($name, $value = null, $namespace = null)
+    public function addChild($name, $value = null, $namespace = null): ?self
```
Comment (ihor-sviziev) [inline app/code/Magento/Shipping/Model/Simplexml/Element.php]:
```suggestion
    #[\ReturnTypeWillChange]
    public function addChild($name, $value = null, $namespace = null)
```

File: app/code/Magento/Shipping/Model/Simplexml/Element.php [hunk 10]
```diff
@@ -20,7 +20,7 @@ class Element extends \Magento\Framework\Simplexml\Element
      * @param string $namespace If specified, the namespace to which the attribute belongs.
      * @return void
      */
-    public function addAttribute($name, $value = null, $namespace = null)
+    public function addAttribute($name, $value = null, $namespace = null): void
```
Comment (ihor-sviziev) [inline app/code/Magento/Shipping/Model/Simplexml/Element.php]:
```suggestion
    #[\ReturnTypeWillChange]
    public function addAttribute($name, $value = null, $namespace = null)
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]

**Review by andrewbess — APPROVED** [review]
Hello @karyna-tsymbal-atwix 
Thank you for your work here.
Approved ✔️

**Review by ihor-sviziev — APPROVED** [review]

File: app/code/Magento/Analytics/Model/ReportXml/ModuleIterator.php (hunk 1 above)
Comment (andrewbess) [reply app/code/Magento/Analytics/Model/ReportXml/ModuleIterator.php]: Hello @ihor-sviziev 
This is really true.
This is big BIC.
We fixed this changes.
@karyna-tsymbal-atwix thank you for the fast fix.

File: app/code/Magento/Analytics/ReportXml/Query.php (hunk 2 above)
Comment (andrewbess) [reply app/code/Magento/Analytics/ReportXml/Query.php]: Hello @ihor-sviziev 
This is really true.
This is big BIC IMHO.
We fixed this changes.
@karyna-tsymbal-atwix thank you for the fast fix.

File: app/code/Magento/Customer/Model/Indexer/Source.php (hunk 3 above)
Comment (andrewbess) [reply app/code/Magento/Customer/Model/Indexer/Source.php]: Hello @ihor-sviziev
This is really true.
This is big BIC IMHO.
We fixed this changes.
@karyna-tsymbal-atwix thank you for the fast fix.

File: app/code/Magento/Customer/Model/Indexer/Source.php (hunk 4 above)
Comment (andrewbess) [reply app/code/Magento/Customer/Model/Indexer/Source.php]: Hello @ihor-sviziev
This is really true.
This is big BIC IMHO.
We fixed this changes.
@karyna-tsymbal-atwix thank you for the fast fix.

File: app/code/Magento/Customer/Ui/Component/MassAction/Group/Options.php (hunk 5 above)
Comment (andrewbess) [reply app/code/Magento/Customer/Ui/Component/MassAction/Group/Options.php]: Hello @ihor-sviziev
This is really true.
This is big BIC IMHO.
We fixed this changes.
@karyna-tsymbal-atwix thank you for the fast fix.

File: app/code/Magento/Ui/Model/Export/SearchResultIterator.php (hunk 6 above)
Comment (andrewbess) [reply app/code/Magento/Ui/Model/Export/SearchResultIterator.php]: Hello @ihor-sviziev
This is really true.
This is big BIC IMHO.
We fixed this changes.
@karyna-tsymbal-atwix thank you for the fast fix.

File: app/code/Magento/Ui/Model/Export/SearchResultIterator.php (hunk 7 above)
Comment (andrewbess) [reply app/code/Magento/Ui/Model/Export/SearchResultIterator.php]: Hello @ihor-sviziev
This is really true.
This is big BIC IMHO.
We fixed this changes.
@karyna-tsymbal-atwix thank you for the fast fix.

File: app/code/Magento/Ui/Model/Export/SearchResultIterator.php (hunk 8 above)
Comment (andrewbess) [reply app/code/Magento/Ui/Model/Export/SearchResultIterator.php]: Hello @ihor-sviziev
This is really true.
This is big BIC IMHO.
We fixed this changes.
@karyna-tsymbal-atwix thank you for the fast fix.

File: app/code/Magento/Shipping/Model/Simplexml/Element.php (hunk 9 above)
Comment (andrewbess) [reply app/code/Magento/Shipping/Model/Simplexml/Element.php]: Hello @ihor-sviziev
This is really true.
This is big BIC IMHO.
We fixed this changes.
@karyna-tsymbal-atwix thank you for the fast fix.

File: app/code/Magento/Shipping/Model/Simplexml/Element.php (hunk 10 above)
Comment (andrewbess) [reply app/code/Magento/Shipping/Model/Simplexml/Element.php]: Hello @ihor-sviziev
This is really true.
This is big BIC IMHO.
We fixed this changes.
@karyna-tsymbal-atwix thank you for the fast fix.

**Comment by andrewbess** [comment]
@magento run Static Tests with env PHP 8.1

**Comment by andrewbess** [comment]
@magento run Static Tests

**Comment by andrewbess** [comment]
Hello @karyna-tsymbal-atwix 
Could you please fix Static Tests?
Thank you in advance.

**Comment by karyna-tsymbal-atwix** [comment]
@magento run Static Tests

**Comment by karyna-tsymbal-atwix** [comment]
@andrewbess  tests were fixed except the "Comment block is missing" warning. This requires to update Magento Coding Standards to ignore #[\ReturnTypeWillChange] after the docblock.

**Comment by karyna-tsymbal-atwix** [comment]
@magento run Static Tests

**Comment by andrewbess** [comment]
@magento run Static Tests, Semantic Version Checker

File: app/code/Magento/Catalog/Model/ResourceModel/Product/Collection/ProductLimitation.php [hunk 11]
```diff
@@ -137,13 +159,15 @@ public function isUsingPriceIndex()
     }
 
     /**
+     * Sets 'Use Price' index flaf.
```
Comment (fascinosum) [inline app/code/Magento/Catalog/Model/ResourceModel/Product/Collection/ProductLimitation.php]: typo

File: app/code/Magento/Ui/DataProvider/AbstractDataProvider.php [hunk 12]
```diff
@@ -162,6 +162,8 @@ public function getFieldMetaInfo($fieldSetName, $fieldName)
 
     /**
      * @inheritdoc
+     * @phpcsSuppress SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingParameterTypeHint
```
Comment (fascinosum) [inline app/code/Magento/Ui/DataProvider/AbstractDataProvider.php]: why do we need this suppress annotation?

File: app/code/Magento/Ui/Model/Export/SearchResultIterator.php [hunk 13]
```diff
@@ -22,40 +22,49 @@ public function __construct(
     }
 
     /**
+     * @inheritDoc
+     *
      * @return array|mixed
```
Comment (fascinosum) [inline app/code/Magento/Ui/Model/Export/SearchResultIterator.php]: if we use `@inheritdoc`, the following `return` annotation is redundant,  especially in the form of `array|mixed`

File: app/code/Magento/Ui/Model/Export/SearchResultIterator.php [hunk 14]
```diff
@@ -22,40 +22,49 @@ public function __construct(
...
      */
+    #[\ReturnTypeWillChange]
     public function current()
     {
         return current($this->items);
     }
 
     /**
+     * @inheritDoc
```
Comment (fascinosum) [inline app/code/Magento/Ui/Model/Export/SearchResultIterator.php]: the same here

**Review by fascinosum — CHANGES_REQUESTED** [review]

File: app/code/Magento/Ui/DataProvider/AbstractDataProvider.php (hunk 12 above)
Comment (xmav) [reply app/code/Magento/Ui/DataProvider/AbstractDataProvider.php:165]: There is return type declared for this. method in interface
```
    /**
     * Add field filter to collection
     *
     * @param \Magento\Framework\Api\Filter $filter
     * @return mixed
     */
    public function addFilter(\Magento\Framework\Api\Filter $filter);
```
So we either should suppress it or update interface. 
Detected by static test
```
2) Magento\Test\Php\LiveCodeTest::testPhpStan
PHPStan detected violation(s):
 ------ ------------------------------------------------------------------ 
  Line   app/code/Magento/Ui/DataProvider/AbstractDataProvider.php         
 ------ ------------------------------------------------------------------ 
  168    Method Magento\Ui\DataProvider\AbstractDataProvider::addFilter()  
         should return mixed but return statement is missing.              
 ------ ------------------------------------------------------------------ 
```
Do you think we should process thought approval process and update interface ? I believe update to annotation in interface will cause SVC failure (It looks like we don't need return type in this method)

File: app/code/Magento/Ui/Model/Export/SearchResultIterator.php (hunk 13 above)
Comment (karyna-t) [reply app/code/Magento/Ui/Model/Export/SearchResultIterator.php]: Thanks, done

File: app/code/Magento/Ui/Model/Export/SearchResultIterator.php (hunk 14 above)
Comment (karyna-t) [reply app/code/Magento/Ui/Model/Export/SearchResultIterator.php]: fixes

File: app/code/Magento/Ui/DataProvider/AbstractDataProvider.php (hunk 12 above)
Comment (karyna-t) [reply app/code/Magento/Ui/DataProvider/AbstractDataProvider.php]: we only need MissingReturnTypeHint , I've removed this extra annotation

**Comment by karyna-tsymbal-atwix** [comment]
@magento run Static Tests

**Comment by karyna-tsymbal-atwix** [comment]
@magento run Static Tests

**Comment by karyna-tsymbal-atwix** [comment]
@magento run Static Tests

**Pull request merged**